File: android/app/src/main/java/tech/acab/app/ble/DetectionNotifier.kt
package tech.acab.app.ble

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import tech.acab.app.MainActivity
import tech.acab.app.R
import tech.acab.app.model.Detection
import tech.acab.app.model.DeviceType
import tech.acab.app.model.displayName
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * Per-category phone notifications for detections. Mirrors iOS DetectionNotifier rule for rule.
 *
 * WHY THIS IS SEPARATE FROM AlertMode: AlertMode governs the BOARD's buzzer (and, in VIBRATE, an
 * in-app haptic). Silencing the board is not the same decision as silencing your phone.
 *
 * ITS OWN CHANNEL, NOT the foreground-service one. AcabLinkService's channel is IMPORTANCE_LOW by
 * design (a persistent status line must never buzz). Detection alerts need HIGH, and the user must
 * be able to tune or mute one without losing the other, which a shared channel forbids.
 *
 * DESIGN RULES:
 *  - Per CATEGORY, opt-in, every one default OFF.
 *  - THE COOLDOWN IS THE EDGE, not the caller's first-sighting flag. The first version gated on
 *    AcabBleManager's `firstTime`, but the detection store is PERSISTED across launches, so a
 *    device seen in any earlier session was never "first" again and could never notify.
 *  - Never for ignored devices (dropped before the hook), never for Desert-mode NEARBY_DEVICE.
 *  - POST_NOTIFICATIONS (API 33+) requested lazily on the first enable.
 *
 * THREADING: the detection path runs on a BINDER thread. BluetoothGattCallback.onCharacteristicChanged
 * -> ingest() -> fileLive() -> here, none of which hops to main. The rate-limit state is therefore
 * concurrent-safe by construction (ConcurrentHashMap + AtomicLong) rather than by assumption; the
 * first version used a plain HashMap and a plain Long, which is a genuine data race next to
 * AcabBleManager's own synchronized(storeLock) neighbours.
 */
class DetectionNotifier(private val ctx: Context) {

    companion object {
        const val CHANNEL_ID = "acab_detections"
        private const val PREFS = "acab"
        private const val KEY_PREFIX = "notify_"

        /** The real rate limit, matching iOS and the settings copy. */
        private const val PER_DEVICE_COOLDOWN_MS = 600_000L    // 10 minutes

        /**
         * Floor between notifications for DIFFERENT devices. Small on purpose: it suppresses one
         * alert, and because the cooldown (not a first-sighting flag) is the edge, the device is
         * offered again on its next sighting rather than lost.
         */
        private const val GLOBAL_MIN_GAP_MS = 4_000L

        /** Prune the cooldown map past this size so a long drive cannot grow it without bound. */
        private const val COOLDOWN_PRUNE_AT = 256

        /**
         * Categories a user can switch on. ONE ALPR row: everywhere else in both apps FLOCK_CAMERA
         * and FLOCK_RAVEN are a single "ALPR" category, and Raven is an acoustic sensor rather than
         * a plate reader, so a separate row subtitled "plate readers" was both a duplicate-looking
         * row and factually wrong. NEARBY_DEVICE is excluded (Desert is ambient, not an event).
         */
        val NOTIFIABLE = listOf(
            DeviceType.FLOCK_CAMERA, DeviceType.BODY_CAM, DeviceType.GLASSES,
            DeviceType.NETWORK_CAMERA, DeviceType.DRONE, DeviceType.TRACKER, DeviceType.WATCHED,
        )

        private fun key(t: DeviceType) = "$KEY_PREFIX${t.raw}"

        /** Cached: the detection path consults this per record, and it used to be up to 8 prefs
         *  reads (each taking a lock) on every one. */
        @Volatile private var enabledCache: Set<Int>? = null

        private fun rebuildCache(ctx: Context) {
            val p = ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            val s = HashSet<Int>()
            NOTIFIABLE.forEach { t ->
                if (p.getBoolean(key(t), false)) {
                    s.add(t.raw)
                    // ALPR is one category; RAVEN has no row and follows FLOCK_CAMERA's switch.
                    if (t == DeviceType.FLOCK_CAMERA) s.add(DeviceType.FLOCK_RAVEN.raw)
                }
            }
            enabledCache = s
        }

        fun isEnabled(ctx: Context, t: DeviceType): Boolean {
            val c = enabledCache ?: run { rebuildCache(ctx); enabledCache }
            return c?.contains(t.raw) == true
        }

        fun setEnabled(ctx: Context, t: DeviceType, on: Boolean) {
            ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
                .putBoolean(key(t), on).apply()
            rebuildCache(ctx)
        }

        fun anyEnabled(ctx: Context): Boolean {
            val c = enabledCache ?: run { rebuildCache(ctx); enabledCache }
            return c?.isNotEmpty() == true
        }

        /** API 33+ gates posting behind a runtime permission; below that it is implicit. */
        fun hasPostPermission(ctx: Context): Boolean =
            Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU ||
                ContextCompat.checkSelfPermission(ctx, Manifest.permission.POST_NOTIFICATIONS) ==
                PackageManager.PERMISSION_GRANTED

        /** True when a category is on but the system will not deliver, so the UI can say so
         *  instead of showing a green toggle over a dead feature. */
        fun mutedBySystem(ctx: Context): Boolean {
            if (!anyEnabled(ctx)) return false
            if (!hasPostPermission(ctx)) return true
            if (!NotificationManagerCompat.from(ctx).areNotificationsEnabled()) return true
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return false
            val nm = ctx.getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
                ?: return false
            val ch = nm.getNotificationChannel(CHANNEL_ID) ?: return false
            return ch.importance == NotificationManager.IMPORTANCE_NONE
        }

        /** True when the Live Mode foreground notification can actually reach the shade:
         *  runtime permission held, app-level notifications on, and the drive channel not
         *  muted. THE single owner of this rule: MainActivity's automatic Live Mode start and
         *  the Beacon readiness surface both read it, so they can never disagree about the
         *  same phone. A null channel passes (it is created on the first service start). */
        fun liveChannelDeliverable(ctx: Context): Boolean {
            if (!hasPostPermission(ctx) ||
                !NotificationManagerCompat.from(ctx).areNotificationsEnabled()) return false
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return true
            val nm = ctx.getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
            return nm?.getNotificationChannel(AcabLinkService.CHANNEL_ID)?.importance !=
                NotificationManager.IMPORTANCE_NONE
        }
    }

    private val lastByMac = ConcurrentHashMap<String, Long>()
    private val lastAt = AtomicLong(0L)

    init {
        // Created up front, not on first post, so the channel exists in system settings for the
        // user to tune BEFORE anything is ever detected.
        ensureChannel()
    }

    private fun ensureChannel() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
        val nm = ctx.getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager ?: return
        if (nm.getNotificationChannel(CHANNEL_ID) != null) return
        nm.createNotificationChannel(
            NotificationChannel(CHANNEL_ID, "Detections", NotificationManager.IMPORTANCE_HIGH).apply {
                description = "Alerts for the detection categories you switched on."
                enableVibration(true)
            }
        )
    }

    /** Same target as the drive-mode notification: open the app on the new-detections log. */
    private fun tapIntent(): PendingIntent = PendingIntent.getActivity(
        ctx, 1,
        Intent(ctx, MainActivity::class.java).putExtra(MainActivity.EXTRA_OPEN_LOG_NEW, true),
        PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT,
    )

    /**
     * Post for a detection if its category is on and the rate limits allow.
     *
     * REDACTION: [redact] is the app's LOCK-SCREEN setting, so it is applied the way Android
     * actually models that, via setPublicVersion, exactly as AcabLinkService.buildPublic already
     * does for drive mode. The first version redacted the main content instead, which hid the
     * category on an UNLOCKED phone too, i.e. in the state the feature exists for. (Redaction
     * itself now defaults OFF; this path only runs for users who turned it on.)
     */
    fun notifyIfNeeded(d: Detection, redact: Boolean) {
        if (d.hist) return                             // an offline replay is not a live event
        if (d.type == DeviceType.NEARBY_DEVICE) return // Desert rows are ambient, never an alert
        if (!isEnabled(ctx, d.type)) return
        if (!hasPostPermission(ctx)) return

        val now = System.currentTimeMillis()
        val prev = lastByMac[d.mac]
        if (prev != null && now - prev < PER_DEVICE_COOLDOWN_MS) return
        // Burst floor for DIFFERENT devices. Safe to drop: the device is not marked, so its next
        // sighting is offered again.
        if (now - lastAt.get() < GLOBAL_MIN_GAP_MS) return

        lastByMac[d.mac] = now
        lastAt.set(now)
        pruneCooldowns(now)

        val who = d.displayName.ifBlank { d.type.label }
        val body = if (who == d.type.label) "Detected nearby, ${d.confidence}% confidence."
                   else "$who detected, ${d.confidence}% confidence."

        val b = NotificationCompat.Builder(ctx, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_stat_beacons)
            .setContentTitle(d.type.label)
            .setContentText(body)
            .setPriority(NotificationCompat.PRIORITY_HIGH)
            // STATUS, not ALARM. CATEGORY_ALARM punches through Do Not Disturb, which contradicts
            // the sibling haptic path that deliberately honours DND via focusSuppressed(), and is
            // not a promise this feature should be making on the user's behalf.
            .setCategory(NotificationCompat.CATEGORY_STATUS)
            .setAutoCancel(true)
            .setContentIntent(tapIntent())
        if (redact) {
            b.setVisibility(NotificationCompat.VISIBILITY_PRIVATE)
            b.setPublicVersion(
                NotificationCompat.Builder(ctx, CHANNEL_ID)
                    .setSmallIcon(R.drawable.ic_stat_beacons)
                    .setContentTitle("beacons")
                    .setContentText("Something was detected nearby.")
                    .setContentIntent(tapIntent())
                    .build()
            )
        } else {
            b.setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
        }
        try {
            // Per-MAC id so a repeat REPLACES rather than stacks. Tagged so it cannot collide with
            // the foreground-service notification's own id space.
            NotificationManagerCompat.from(ctx).notify("acab.det", d.mac.hashCode(), b.build())
        } catch (_: SecurityException) {
            // Permission revoked between the check above and the post. The next enable re-requests.
        }
    }

    private fun pruneCooldowns(now: Long) {
        if (lastByMac.size <= COOLDOWN_PRUNE_AT) return
        lastByMac.entries.removeAll { now - it.value >= PER_DEVICE_COOLDOWN_MS }
    }

    /** Drop cooldown state so a genuinely new session can alert on the same devices again. */
    fun reset() {
        lastByMac.clear()
        lastAt.set(0L)
    }
}

File: ios/Beacons/BLE/DetectionNotifier.swift
import Foundation
import UserNotifications

/// Per-category phone notifications for detections.
///
/// WHY THIS IS SEPARATE FROM AlertMode: `AlertMode` governs the BOARD's buzzer (and, in `.vibrate`,
/// an in-app haptic that only fires while the app is foregrounded). It is not a notification
/// system, and before this file the app had none at all. Silencing the board is not the same
/// decision as silencing your phone, so the two are deliberately independent axes.
///
/// DESIGN RULES, each of which exists because the alternative is a firehose or a lie:
///  - Per CATEGORY, opt-in, every one default OFF.
///  - THE COOLDOWN IS THE EDGE, not the caller's `firstTime`. This is the correction that matters:
///    the first version gated on BLEManager's `firstTime`, which is `store[d.id] == nil`. The store
///    is PERSISTED across launches, so a device seen in any previous session was never "first"
///    again and could never notify, and the ten-minute cooldown decided nothing. Worse, a detection
///    suppressed by the global gap was destroyed rather than deferred, because there was only ever
///    one offer per device. Owning the dedup here makes both correct and matches the shipped copy.
///  - Never for ignored devices (dropped before the hook), never for Desert-mode `.nearbyDevice`.
///  - Authorization is requested LAZILY on the first enable, not at launch.
///
/// FOREGROUND PRESENTATION: this must be the notification-center delegate. Without one, iOS
/// silently discards the alert whenever the app is on screen: no banner, no sound, no entry, and
/// `add()` reports success. Since the app has no in-app banner either, the feature produced
/// literally nothing while you were looking at it.
///
/// THREADING: plain NSObject, no actor isolation, because BLEManager is a non-isolated
/// NSObject/ObservableObject and a @MainActor property initializer cannot be constructed from it.
/// Entry points run on the main thread (CoreBluetooth is created with a nil queue, so its callbacks
/// are main) and the async completion handlers hop back to main explicitly.
final class DetectionNotifier: NSObject, UNUserNotificationCenterDelegate {

    /// Per-device cooldown. This is the real rate limit, and the only one: a board re-reports a
    /// device it can still hear about once a second, so without this a single camera would notify
    /// continuously.
    private static let perDeviceCooldown: TimeInterval = 600      // 10 minutes, matches the UI copy

    /// Floor between notifications for DIFFERENT devices, so arriving somewhere dense cannot produce
    /// a burst you cannot read. Deliberately small: it suppresses one alert, and because the
    /// cooldown (not `firstTime`) is now the edge, the device gets another chance on its next
    /// sighting a second later rather than being silently dropped forever.
    private static let globalMinGap: TimeInterval = 4

    private var lastNotifiedByMac: [String: Date] = [:]
    private var lastNotifiedAt: Date = .distantPast
    private(set) var authorized = false
    private var authRequested = false

    /// Cached because the detection path consults it per record. It used to be recomputed from 8
    /// UserDefaults reads on every one, which is hot enough to matter at Desert-mode rates.
    private static var enabledCache: Set<Int>?

    override init() {
        super.init()
        // Must be set before any notification is posted, and this object is built during
        // BLEManager's own construction, which happens at app launch.
        UNUserNotificationCenter.current().delegate = self
    }

    // MARK: - Foreground presentation

    /// Show the banner even when the app is foregrounded. Without this the alert is dropped
    /// entirely. The sound is withheld when the user has chosen a silent board and a Focus is
    /// active, matching what the in-app haptic already does rather than chirping over it.
    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                willPresent notification: UNNotification,
                                withCompletionHandler completionHandler:
                                @escaping (UNNotificationPresentationOptions) -> Void) {
        completionHandler(silenceForeground ? [.banner, .list] : [.banner, .list, .sound])
    }

    /// Set by BLEManager so the delegate can honor the same Focus/alert-mode rule the haptic uses.
    var silenceForeground = false

    /// Tapping the notification opens the Log filtered to new detections, the same destination
    /// Android's contentIntent uses and the same one the offline-sync banner's "view" action
    /// already drives. Without this the notification was inert on tap, which on iOS just
    /// foregrounds the app on whatever screen it was last on.
    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                didReceive response: UNNotificationResponse,
                                withCompletionHandler completionHandler: @escaping () -> Void) {
        // Deferred one runloop turn ON PURPOSE. RootView.onAppear clears both pending keys once
        // per process, deliberately, so a stale tap cannot replay days later. `.onOpenURL` is
        // documented to land after that clear; a notification response has no such guarantee on a
        // COLD launch, and setting the keys first would let onAppear wipe them. Hopping to the
        // next turn puts us after any synchronous onAppear either way.
        DispatchQueue.main.async {
            UserDefaults.standard.set(true, forKey: "acab.pendingNewFilter")
            UserDefaults.standard.set(2, forKey: "acab.pendingTab")
            NotificationCenter.default.post(name: Notification.Name("acabOpenLogNew"), object: nil)
        }
        completionHandler()
    }

    // MARK: - Preferences

    private static func key(for type: DeviceType) -> String { "acab.notify.\(type.rawValue)" }

    static func isEnabled(_ type: DeviceType) -> Bool {
        if enabledCache == nil { rebuildCache() }
        return enabledCache?.contains(type.rawValue) ?? false
    }

    private static func rebuildCache() {
        var s = Set<Int>()
        for t in notifiableTypes where UserDefaults.standard.bool(forKey: key(for: t)) {
            s.insert(t.rawValue)
            // ALPR is one user-facing category. flockRaven has no row of its own (see
            // notifiableTypes), so it follows flockCamera's switch.
            if t == .flockCamera { s.insert(DeviceType.flockRaven.rawValue) }
        }
        enabledCache = s
    }

    /// The categories a user can switch on. ONE ALPR row: everywhere else in both apps flockCamera
    /// and flockRaven are a single "ALPR" category, and Raven is an acoustic sensor rather than a
    /// plate reader, so listing it separately under "plate readers" was both a duplicate-looking
    /// row and factually wrong. `.nearbyDevice` is excluded (Desert is ambient, not an event).
    static let notifiableTypes: [DeviceType] = [
        .flockCamera, .axonBodyCam, .recordingGlasses,
        .networkCamera, .drone, .tracker, .watched,
    ]

    static var anyEnabled: Bool {
        if enabledCache == nil { rebuildCache() }
        return !(enabledCache?.isEmpty ?? true)
    }

    func setEnabled(_ on: Bool, for type: DeviceType) {
        UserDefaults.standard.set(on, forKey: Self.key(for: type))
        Self.rebuildCache()
        if on { requestAuthorizationIfNeeded() }
    }

    // MARK: - Authorization

    func requestAuthorizationIfNeeded() {
        guard !authRequested else { return }
        authRequested = true
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound, .badge]) {
            [weak self] granted, _ in
            DispatchQueue.main.async { self?.authorized = granted }
        }
    }

    /// Re-read the real system state. Called on launch AND on every foreground: a user who grants
    /// permission in iOS Settings mid-session would otherwise stay silently unauthorized for the
    /// whole process lifetime, since `authorized` is only ever set by our own request.
    ///
    /// `completion` runs on the main queue AFTER the cached state is written, so a caller that
    /// re-renders off `mutedBySystem` can bump its own invalidation token there and be guaranteed
    /// to repaint against the fresh answer, not the stale cache.
    func refreshAuthorization(completion: (() -> Void)? = nil) {
        UNUserNotificationCenter.current().getNotificationSettings { [weak self] s in
            DispatchQueue.main.async {
                self?.authorized = (s.authorizationStatus == .authorized ||
                                    s.authorizationStatus == .provisional)
                // `authRequested` means "iOS has already answered", NOT "we were granted". Setting
                // it only on success meant a DENIED user came back false on every launch, so
                // mutedBySystem stayed false and the amber "iOS is blocking these" warning could
                // never appear again: green toggles over a permanently dead feature, which is the
                // exact outcome that warning exists to prevent.
                if s.authorizationStatus != .notDetermined { self?.authRequested = true }
                completion?()
            }
        }
    }

    /// True when the user has switched something on but the system will not deliver it, so the UI
    /// can say so instead of showing a green toggle over a dead feature.
    var mutedBySystem: Bool { Self.anyEnabled && !authorized && authRequested }

    // MARK: - Posting

    /// Post for a detection if its category is on and the rate limits allow.
    ///
    /// NOT gated on the caller's first-sighting edge, deliberately. See the type comment.
    ///
    /// REDACTION: the notification TEXT is never redacted here. `redactLockScreen` is a
    /// lock-screen-scoped setting, and redacting the body unconditionally hid the category on an
    /// UNLOCKED phone too, which is the state the feature exists for. iOS already has the right
    /// mechanism: the user's own "Show Previews" setting governs whether the body appears on a
    /// locked screen, per-app and system-wide. Honoring that instead of second-guessing it is both
    /// correct and less surprising.
    func notifyIfNeeded(_ d: Detection) {
        guard authorized,
              !d.isHistory,                       // an offline replay is not a live event
              d.type != .nearbyDevice,            // Desert rows are ambient, never an alert
              Self.isEnabled(d.type)
        else { return }

        let now = Date()
        if let last = lastNotifiedByMac[d.mac], now.timeIntervalSince(last) < Self.perDeviceCooldown {
            return
        }
        // Different-device burst floor. Suppressing here is safe now: the device is not marked, so
        // its next sighting is offered again rather than lost.
        guard now.timeIntervalSince(lastNotifiedAt) >= Self.globalMinGap else { return }

        lastNotifiedByMac[d.mac] = now
        lastNotifiedAt = now
        pruneCooldowns(now)

        let content = UNMutableNotificationContent()
        content.title = d.type.label
        let who = d.hasName ? d.displayName : d.type.label
        // Avoid "Tracker / Tracker detected" when the device has no name of its own.
        content.body = who == d.type.label
            ? "Detected nearby, \(d.confidence)% confidence."
            : "\(who) detected, \(d.confidence)% confidence."
        content.sound = .default

        // NOTE: .timeSensitive is deliberately NOT set. It requires the
        // com.apple.developer.usernotifications.time-sensitive entitlement, which this target does
        // not declare, so iOS silently downgrades it and the line reads like a working feature
        // while doing nothing. If Focus breakthrough is wanted, add the entitlement to
        // project.yml AND the provisioning profile first, then set it here. Until then this
        // matches Android, which no longer uses CATEGORY_ALARM for the same reason.

        // Per-MAC identifier so a repeat replaces rather than stacks.
        let req = UNNotificationRequest(identifier: "acab.det.\(d.mac)", content: content, trigger: nil)
        UNUserNotificationCenter.current().add(req) { err in
            if let err { print("[ACAB] notification post failed: \(err.localizedDescription)") }
        }
    }

    /// Drop expired entries so a long drive past thousands of devices cannot grow this without
    /// bound. Cheap: it only runs when something was actually posted.
    private func pruneCooldowns(_ now: Date) {
        guard lastNotifiedByMac.count > 256 else { return }
        lastNotifiedByMac = lastNotifiedByMac.filter {
            now.timeIntervalSince($0.value) < Self.perDeviceCooldown
        }
    }

    /// Drop cooldown state so a genuinely new session can alert on the same devices again.
    func reset() {
        lastNotifiedByMac.removeAll()
        lastNotifiedAt = .distantPast
    }
}
